### PR TITLE
Backport CMake configuration fix for test_allocator

### PR DESCRIPTION
**Pull request by hobu**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

This applies the patch found in https://github.com/microsoft/vcpkg/pull/39664 to get abseil compiling again.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -53,22 +54,19 @@ outputs:
...
-        - __osx >={{ MACOSX_DEPLOYMENT_TARGET }}  # [osx and x86_64]
 
     requirements:
       build:
         - {{ compiler('c') }}
         - {{ compiler('cxx') }}
+        - {{ stdlib('c') }}
         - cmake
         - ninja
+        - gtest
```
Comment (xhochy) [inline recipe/meta.yaml:65]: This should be in host.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml (hunk 1 above)
Comment (hobu) [reply recipe/meta.yaml:66]: addressed

File: recipe/meta.yaml [hunk 2]
```diff
@@ -37,9 +37,10 @@ source:
     - patches/0002-don-t-use-C-20-stdlib-features-which-change-ABI-comp.patch
     # workaround for https://github.com/abseil/abseil-cpp/issues/1624
     - patches/0003-unconditionally-export-Mutex-Destructor.patch
+    - patches/0004-test-allocator.patch
```
Comment (h-vetinari) [inline recipe/meta.yaml:40]: Please add proper context and attribution for the patch
```suggestion
    # backport https://github.com/abseil/abseil-cpp/commit/779a3565ac6c5b69dd1ab9183e500a27633117d5
    - patches/<will-be-renamed,-see-below>.patch
```

File: recipe/patches/0004-test-allocator.patch [hunk 3]
```diff
@@ -0,0 +1,12 @@
+diff --git a/absl/container/CMakeLists.txt b/absl/container/CMakeLists.txt
+index 449a2cad17e..ee9ca9c3c4a 100644
```
Comment (h-vetinari) [inline recipe/patches/0004-test-allocator.patch:2]: Please generate this patch from https://github.com/abseil/abseil-cpp/commit/779a3565ac6c5b69dd1ab9183e500a27633117d5 using `git format-patch`.

If you want, here are more complete instructions:
```
git clone https://github.com/abseil/abseil-cpp
cd abseil-cpp
git checkout tags/20240116.2
find ../path/to/abseil-cpp-feedstock/recipe/patches/ | xargs git am      # without your 0004-* in recipe/patches
git cherry-pick 779a3565ac6c5b69dd1ab9183e500a27633117d5
git format-patch tags/20240116.2 --no-signature
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -53,22 +54,20 @@ outputs:
         - {{ pin_subpackage("libabseil", max_pin="x") }}
         # also pin on ABI variant
         - libabseil =*=cxx{{ cxx_standard }}*
-        # abseil needs the 10.13 headers even just to be `# include`d;
-        # ensure that all dependent packages reflect that to
-        # avoid the solver getting stuck in no-man's-land.
-        - __osx >={{ MACOSX_DEPLOYMENT_TARGET }}  # [osx and x86_64]
```
Comment (h-vetinari) [inline recipe/meta.yaml:57]: we don't want the run-export from gtest for libabseil itself
```suggestion
      ignore_run_exports:
        - gtest

```

**Review by h-vetinari — CHANGES_REQUESTED** [review]
This PR is titled misleadingly - the stdlib stuff is routine and not necessary to pull forward from whenever the next version bump would occur, but what you're after IIUC is to apply a patch "to get abseil compiling again" (<- what are the conditions where something breaks?), which is what should be indicated in the title.

File: recipe/meta.yaml (hunk 2 above)
Comment (hobu) [reply recipe/meta.yaml]: updated as requested

File: recipe/meta.yaml (hunk 4 above)
Comment (hobu) [reply recipe/meta.yaml:60]: Updated as requested

**Comment by hobu** [comment]
> This PR is titled misleadingly - the stdlib stuff is routine and not necessary to pull forward from whenever the next version bump would occur, but what you're after IIUC is to apply a patch "to get abseil compiling again" (<- what are the conditions where something breaks?), which is what should be indicated in the title.

Indeed. Without the patch, the feedstock does not render for me on osx-arm64 similarly to the ticket that was referenced for vcpkg:

```
CMake Error at CMake/AbseilHelpers.cmake:317 (target_link_libraries):
  The link interface of target "test_allocator" contains:

    GTest::gmock

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  absl/container/CMakeLists.txt:206 (absl_cc_library)

```

File: recipe/patches/0004-test-allocator.patch (hunk 3 above)
Comment (hobu) [reply recipe/patches/0004-test-allocator.patch]: Thanks for the instructions. They were very helpful, and I have updated the patch as requested.

**Comment by h-vetinari** [comment]
> Without the patch, the feedstock does not render for me on osx-arm64

Can you clarify this a bit -- "(re)rendering" is what we use to describe our bots updating the feedstock infrastructure (all the stuff not under `/recipe`), so I presume you don't mean that. Are you trying to build the feedstock locally? Or do you mean compiling something else locally that requires abseil?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -53,22 +55,22 @@ outputs:
...
 
     requirements:
       build:
         - {{ compiler('c') }}
         - {{ compiler('cxx') }}
+        - {{ stdlib('c') }}
         - cmake
         - ninja
+      host:
+        - gtest
```
Comment (njzjz) [inline recipe/meta.yaml:69]: I understand that you want to patch https://github.com/abseil/abseil-cpp/commit/779a3565ac6c5b69dd1ab9183e500a27633117d5, but is there any reason to add gtest as the dependency?

Comment (h-vetinari) [reply recipe/meta.yaml:69]: That's what I've been trying to figure out as well. At least, with the `ignore_run_exports`, it wouldn't affect the artefacts we produce (in terms of run dependencies).

Comment (hobu) [reply recipe/meta.yaml:69]: It is not needed, and I have since removed it. The compiler error above was a red herring.

**Comment by hobu** [comment]
> Are you trying to build the feedstock locally? Or do you mean compiling something else locally that requires abseil?

I'm trying to compile the feedstock locally with `conda build`

> `conda mambabuild recipe -m .ci_support/osx_arm64_.yaml`

It was a yak shaving adventure precipitated by trying to stage a recipe for [protobuf-c](https://github.com/protobuf-c/protobuf-c) and finding the abseil package would not solve for it.

**Comment by hobu** [comment]
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=976058&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=7c0f8eae-6d6f-51bf-636f-73a1a7fb1bc4&l=1867 is also an example of the error that happens if a build with this PR is not available

**Comment by h-vetinari** [comment]
OK, this looks good now in principle, but I'd like to avoid the git history churn of adding things only to remove them again. I can either offer to squash-merge (my less preferred option), or you could do an interactive rebase to condense things down to ~3 commits, e.g.:
* handle stdlib things (only those under `recipe/`)
* add patch
* rerender

The build number increment can be separate or go into either of the first two commits.

File: recipe/meta.yaml [hunk 6]
```diff
@@ -37,6 +37,8 @@ source:
     - patches/0002-don-t-use-C-20-stdlib-features-which-change-ABI-comp.patch
     # workaround for https://github.com/abseil/abseil-cpp/issues/1624
     - patches/0003-unconditionally-export-Mutex-Destructor.patch
+    # backport https://github.com/abseil/abseil-cpp/commit/779a3565ac6c5b69dd1ab9183e500a27633117d5
+    - patches/0004-Avoid-export-of-testonly-target-absl-test_allocator-.patch
 
 build:
   number: 0
```
Comment (h-vetinari) [inline recipe/meta.yaml:44]:
```suggestion
  number: 1
```

**Review by h-vetinari — APPROVED** [review]
Thanks for your patience on this one!

**Pull request merged**